pkg/tenancy/grpc.go

```diff
@@ -22,34 +23,65 @@ func (tss *tenantedServerStream) Context() context.Context {
 	return tss.context
 }
 
-func getValidTenant(ctx context.Context, tc *Manager) (string, error) {
-	// Handle case where tenant is already directly in the context
-	tenant := GetTenant(ctx)
-	if tenant != "" {
-		if !tc.Valid(tenant) {
-			return tenant, status.Errorf(codes.PermissionDenied, "unknown tenant")
+func GetValidTenant(ctx context.Context, tm *Manager) (string, error) {
+	// Check if context directly has tenant attached
+	// Check for OTEL client' metadata
+	// Check for gRPC metadata
+	var tenant string
+	var err error
+	if tenant = GetTenant(ctx); tenant != "" {
+		if !tm.Valid(tenant) {
```

Comment:
please refactor such that you don't repeat logic. You can have a separate helper to extract the tenants (with early returns), and here validate it once.